src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlStringMethodTranslator.cs

```diff
@@ -224,50 +240,79 @@ SqlExpression TranslateStartsEndsWith(SqlExpression instance, SqlExpression patt
                     _sqlExpressionFactory.Equal(
-                        _sqlExpressionFactory.Function(
-                            "LEFT",
-                            new[] {
-                                instance,
-                                _sqlExpressionFactory.Function("LENGTH", new[] { pattern }, typeof(int))
-                            },
-                            typeof(string),
-                            stringTypeMapping),
-                        pattern));
+                        instance.TypeMapping == _textTypeMapping
+                            ? (SqlExpression)leftPart
+                            // We have Convert here since LEFT over citext returns text - cast it back.
+                            : _sqlExpressionFactory.Convert(
+                                leftPart,
+                                typeof(string),
+                                instance.TypeMapping),
+                        // The following Convert is only needed because of https://github.com/aspnet/EntityFrameworkCore/issues/19120
+                        pattern.TypeMapping == _textTypeMapping
+                            ? pattern
+                            : _sqlExpressionFactory.Convert(
+                                pattern,
+                                typeof(string),
+                                pattern.TypeMapping)));
             }
 
+            var rightPart = _sqlExpressionFactory.Function(
+                "RIGHT",
+                new[]
+                {
+                    instance,
+                    _sqlExpressionFactory.Function("LENGTH", new[] { pattern }, typeof(int))
+                },
+                typeof(string),
+                stringTypeMapping);
+
             return _sqlExpressionFactory.Equal(
-                _sqlExpressionFactory.Function(
-                    "RIGHT",
-                    new[] {
-                        instance,
-                        _sqlExpressionFactory.Function("LENGTH", new[] { pattern }, typeof(int))
-                    },
-                    typeof(string),
-                    stringTypeMapping),
-                pattern);
+                instance.TypeMapping == _textTypeMapping
+                    ? (SqlExpression)rightPart
+                    // We have Convert here since RIGHT over citext returns text - cast it back.
+                    : _sqlExpressionFactory.Convert(
+                        rightPart,
+                        typeof(string),
+                        instance.TypeMapping),
+                // The following Convert is only needed because of https://github.com/aspnet/EntityFrameworkCore/issues/19120
+                pattern.TypeMapping == _textTypeMapping
+                    ? pattern
+                    : _sqlExpressionFactory.Convert(
+                        pattern,
+                        typeof(string),
+                        pattern.TypeMapping));
```

Comment:
These two parameters could be generated via two nested methods, and you won't need to duplicate the code.

Reply:
Is it really worth it? I like nested methods for more complicated cases, but here it seems more readable to just keep the code inline, so that everything is clear when reading and you don't have to jump to another function...

Reply:
It's worth at least for conversion to `citext`. It minimizes a chance to make a bug or a fix in the code and not to affect another place.

Reply:
This isn't just about conversion to citext, it could be another text type as well (e.g. json). Can you post a sample of exactly what you mean?

IMHO we shouldn't always, everywhere extract duplicated code into methods - methods make it more difficult to follow what's happening (you have to jump around). It's always a case-by-case decision based on how much is being duplicated, etc. etc.

Reply:
```csharp
// Could be used for RIGHT and LEFT
return _sqlExpressionFactory.Equal(
    GetInstanceMapping(rightPart),
    GetPatternMapping());

SqlExpression GetInstanceMapping(SqlExpression value)
    => instance.TypeMapping == _textTypeMapping
        ? value
        : _sqlExpressionFactory.Convert(value, typeof(string), _textTypeMapping);

// The same method for pattern or make a common method until the bug isn't fixed.
```

Reply:
OK, I ended up refactoring in a slightly different way but I think you'll like it - can you take another look?

Reply:
Now it looks much better. Let's merge (: